UP-4438: Created '/layout/v2/layout.json' endpoint that returns the layo...

groybal opened the pull request:
...ut data needed for rendering the uPortal UI using Javascript.  Also, added '/layout/v1/layout.json' endpoint, which returns the same JSON as the existing '/layout.json' endpoint'.

See:  https://issues.jasig.org/browse/UP-4438
Sample JSON output:  
https://issues.jasig.org/secure/attachment/14744/layout-json-v2-admin-sample.txt
https://issues.jasig.org/secure/attachment/14744/layout-json-v2-student-sample.txt

jameswennmacher commented:
The output looks great.  A JS UI has all the info it needs to construct whatever portions of the page it needs to (we plan on only tab/content in the expected  scenario, but good to have all the regions for the future).   The only thing that occurred to me this time looking at the output is to change 'channel' to 'portlet' to update it to more modern terminology.  Not sure if this is possible but if so, it would be IMHO a bit nicer.

timlevett commented:
Looks good to me. :+1:

jameswennmacher commented on uportal-war/src/main/java/org/jasig/portal/json/rendering/JsonStructureAttributeSource.java:
```diff
@@ -21,24 +21,53 @@
+ * Note that the only difference in the rendering pipeline between the original endpoint and the new versioned endpoints are the stylesheets used.  Because of the number of 
+ * beans required to setup the pipeline (due to the deep chaining of {@link PipelineComponents}), it is desireable to reuse the Spring Framework configuration that sets this 
+ * up rather than duplicating it for each endpoint.  Using a request attribute to specify the stylesheet (instead of using a configured value) allows us to do this 
+ * as each endpoint class or method can specify a different stylesheet.
```
Comment width too large (172+).  Please reformat to 100 per future Google code guidelines (https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.4-column-limit).

jameswennmacher commented on uportal-war/src/main/java/org/jasig/portal/json/rendering/JsonStructureAttributeSource.java:
```diff
@@ -21,24 +21,53 @@
+    }
+
+    protected String getStyleSheetName(final HttpServletRequest request) {
+        final String stylesheetNameFromRequest = (String)request.getAttribute(STYLESHEET_NAME_REQUEST_ATTRIBUTE);
```
Here and above (and elsewhere) please limit lines to 100 columns.

jameswennmacher commented on uportal-war/src/main/java/org/jasig/portal/json/rendering/JsonStructureAttributeSource.java:
```diff
@@ -21,24 +21,53 @@
+        return this.stylesheetDescriptorDao.getStylesheetDescriptorByName(this.getStyleSheetName(request));
+    }
+
+    protected String getStyleSheetName(final HttpServletRequest request) {
```
Add since 4.2.0?

jameswennmacher commented on uportal-war/src/main/java/org/jasig/portal/json/rendering/LayoutJsonV2RenderingController.java:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * @author Eric Dalquist
+ * @version $Revision$
```
Suggest adding @since 4.2.0.  Also suggest adding comment describing what this output is for.

jameswennmacher commented on uportal-war/src/main/resources/layout/structure/columns/columns-imports.xsl:
```diff
@@ -0,0 +1,216 @@
+    specific language governing permissions and limitations
+    under the License.
+
+-->
```
Suggest adding comment about what file is for

jameswennmacher commented on uportal-war/src/main/resources/layout/structure/columns/columns-js.xsl:
```diff
@@ -0,0 +1,184 @@
+    specific language governing permissions and limitations
+    under the License.
+
+-->
```
Suggest adding comment about what file does and how used.

jameswennmacher commented on uportal-war/src/main/resources/layout/theme/json/json-v2.xsl:
```diff
@@ -0,0 +1,378 @@
+ | RED: Stop! Do not modify.
+ | YELLOW: Warning, proceed with caution.  Modifications can be made, but should not generally be necessary and may require more advanced skill.
+ | GREEN: Go! Modify as desired.
+-->
```
Good comment.  Just line length issue.

jameswennmacher commented on uportal-war/src/main/resources/layout/theme/json/json-v2.xsl:
```diff
@@ -0,0 +1,378 @@
+<!-- 
+ | Date: 04/01/2015
+ | Company: Unicon, Inc.
+ | uPortal Version: TBD (uP5.0 ?)
```
uPortal 4.2.0

jameswennmacher commented on uportal-war/src/test/java/org/jasig/portal/json/rendering/JsonStructureAttributeSourceTest.java:
```diff
@@ -0,0 +1,81 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
```
Excellent!  Glad to see unit tests!

jameswennmacher commented:
+1.  Would love to see in uPortal 4.2.0, even though primary usage will be in uPortal 5.0 with richer JS UI.  Would be available for uMobile as well.

jameswennmacher commented:
What do others think?  Leave channel element in the layout.json or change element name to portlet?

groybal commented:
I have made updates per the review comments.  I also went ahead and renamede the "channels" and "channel" elements as "portlets" and "portlet".

jameswennmacher commented:
Created issue https://issues.jasig.org/browse/UP-4444 which shouldn't stop this PR from being processed.

jameswennmacher commented:
I noticed URLs in v2 json are:
"url": "/uPortal/f/u21l1s5/p/notification-icon.u21l1n6/detached/render.uP?pCm=view",

Shouldn't URL should be /exlusive/ not /detached/ ?

jameswennmacher commented:
Noticed windowState is often "minimized" instead of "normal"? It should be normal.  See example output of student user on v2 at https://issues.jasig.org/secure/attachment/14743/new-student-v2-layout.json. Search for lines with JNW>.  I noted one instance but if you search the file you will see many.

jameswennmacher commented:
"hasColumnAddChildAllowed": "true",

Could this attribute be named 'addChildAllowed' to match the name in the fragment-layout files?
Right now it is the only inconsistent attribute name.  See 'layout/v2/layout.json' column added to folders table at https://wiki.jasig.org/display/UPM41/Setup+a+fragment+layout+by+modifying+fragment-layout+XML+files.

Search https://issues.jasig.org/secure/attachment/14743/new-student-v2-layout.json (student user) for JNW> to see instance where this attribute is present.

jameswennmacher commented:
Minor:  hasFavorites is a global value and is present in the debug section (which should probably be named something other than debug, perhaps 'globals' instead).  I suggest removing it from each tab folder's section if that's easy.  In some ways it is not needed at all ... it's value is true if there is a 'favorites'  item in the layout, but I favor leaving hasFavorites in the 'globals' area.

groybal commented:
James, I fixed the 'exclusive' URL window state issue, changed 'hasColumnAddChildAllowed' to 'addChildAllowed', removed 'hasFavorites' from the tabs, and renamed the 'debug' section as 'globals'.  In my testing I did not see the issue with windowState being 'minimized' (only one or two portlets had this state).

jameswennmacher commented:
Thanks Gary.  I was able to verify that the windowState is not an issue at least today.  I'm not sure why I saw it so much earlier but I didn't see it now.  I found that notification-icon portlet shows up as minimized but I don't think that's an issue.

jameswennmacher commented:
I did find one issue needing resolved before we accept this pull request.  Whenever you invoke /uPortal/layout/v2/layout.json, a large number of warnings and stack traces appear in portal/portal.log.  It looks like 4 warnings per portlet.  See https://gist.github.com/jameswennmacher/db476fce629bef8df00c for the warnings.  It looks like after the layout is loaded that an attempt is made to actually render all the portlets.  PortletRenderExecutionWorker and PortletExecutionManager seem to emit 2 warnings each per portlet.

timlevett commented:
Does this happen for v1 as well @jameswennmacher ?

jameswennmacher commented:
@timlevett excellent question.  No the stack traces are specific to v2.

jameswennmacher commented:
@groybal have the stack trace issues been resolved?

jameswennmacher commented:
Did you add comments to columns-imports.xsl and columns-js.xsl describing what the file is for?

jameswennmacher commented:
Andrew S found a bug. If a layout file has a folder with portlets in it (which is allowed), but the portlets are not in a column, the resulting JSON does not show the result.  I suggest making the JSON look like the portlets are in a single column.

drewwills commented on uportal-war/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -414,6 +414,8 @@
     <servlet-mapping>
         <servlet-name>JsonRenderingDispatcherServlet</servlet-name>
+        <url-pattern>/layout/v2/layout.json</url-pattern>
+        <url-pattern>/layout/v1/layout.json</url-pattern>
```
I think the next version (and subsequent versions) of this API should simply be moved under /api so that we don't have to keep managing servlet-mappings in web.xml.

jameswennmacher commented:
Discussed with @drewwills . Thinking of making the API endpoint /uPortal/api/v4-3/dlm/layout.json

jameswennmacher commented:
Replaced with https://github.com/Jasig/uPortal/pull/571
